Test support module (#2357)
spring-cloud/spring-cloud-gcp: docs/src/main/asciidoc/test.adoc

== Emulators support

> Our tools simplify starting and stopping Cloud PubSub and Cloud Spanner emulators when you test your application.

Reviewer: ```suggestion
Our tools simplify starting and stopping Cloud Pub/Sub and Cloud Spanner emulators when you test your applications locally.
```

In order to use it, you need to add this dependency into your `pom.xml` file:

[source,xml]
----
<dependency>
	<groupId>org.springframework.cloud</groupId>
	<artifactId>spring-cloud-gcp-test</artifactId>
	<scope>test</scope>
</dependency>
----

> Also, you need to have `gcloud` and the emulators installed and configured, as described later.

Reviewer: ```suggestion
Also, you need to have `gcloud` CLI and the emulators installed and configured, as described later.
```

=== JUnit 4 Class Rule

> The rule starts the emulator process before the tests run and kills it afterwards.

Reviewer: ```suggestion
The rule starts the emulator process before the tests run and stops it afterwards.
```

You just need to add a rule into your test class to enable it.

[source,java]
----
import org.springframework.cloud.gcp.test.EmulatorRule;
import org.springframework.cloud.gcp.test.pubsub.PubSubEmulator;

public class PubSubTests {
	@ClassRule
	public static EmulatorRule emulator = new EmulatorRule(new PubSubEmulator());

		//your tests
}
----

> NOTE: The class rule doesn't work for SpannerEmulator.

Reviewer: ```suggestion
NOTE: The class rule doesn't work for `SpannerEmulator`.
```

> NOTE: The class rule doesn't work for `SpannerEmulator`.

Author: It looks like implementation details to me. @meltsufin what do you think?

See <<Spring Configuration>> section instead.

=== Utility class
If you prefer controlling the emulator manually, you can use the `EmulatorDriver`.

[source,java]
----
	EmulatorDriver emulatorDriver = new EmulatorDriver(new PubSubEmulator());

	emulatorDriver.startEmulator();

	//your code

	emulatorDriver.shutdownEmulator();
----

> === Cloud PubSub Emulator

Reviewer: ```suggestion
=== Cloud Pub/Sub Emulator
```

> In order to use our Cloud PubSub Emulator helper tools, you would need to install the emulator first.

Reviewer: ```suggestion
In order to use our Cloud Pub/Sub Emulator helper tools, you would need to install the emulator first.
```

Follow the https://cloud.google.com/pubsub/docs/emulator[installation instructions].

> Use `new PubSubEmulator()` as an argument for `EmulatorDriver` and `EmulatorRule`.

Reviewer: ```suggestion
Use `new PubSubEmulator()` as an argument to `EmulatorDriver` and `EmulatorRule`.
```

=== Cloud Spanner Emulator
In order to use our Cloud Spanner Emulator helper tools, you would need to install the emulator first.
Follow the https://cloud.google.com/spanner/docs/emulator[installation instructions].
Make sure you also create an https://cloud.google.com/spanner/docs/emulator#using_the_gcloud_cli_with_the_emulator[emulator configuration] and call it `emulator`.

> Use `new SpannerEmulator()` as an argument for `EmulatorDriver`

Reviewer: ```suggestion
Use `new SpannerEmulator()` as an argument to `EmulatorDriver`.
```

> ==== Spring Configuration

Reviewer: ```suggestion
==== Spanner Emulator Spring Configuration
```

If you are testing your Spring application, you can use our configuration class.

> It provides a bean of type Spanner, which starts the emulator before creating a client.

Reviewer: ```suggestion
It provides a bean of type `Spanner`, which starts the emulator before creating a client.
```
